Apps/Src/SendCarCAN.c

```diff
@@ -1,36 +1,187 @@
  * 
  */
 
-#include "Tasks.h"
-#include "CANbus.h"
-#include "CAN_Queue.h"
 #include "common.h"
+#include "os_cfg_app.h"
+#include "CANbus.h"
+#include "Minions.h"
+#include "Contactors.h"
+#include "Pedals.h"
+#include "Tasks.h"
+#include "SendCarCAN.h"
+#include "SendTritium.h"
+
+#define IO_STATE_DLY_MS 250u 
+
+// Task_PutIOState
+OS_TCB putIOState_TCB;
+CPU_STK putIOState_Stk[TASK_SEND_CAR_CAN_STACK_SIZE];
+
+//fifo
+#define FIFO_TYPE CANDATA_t
+#define FIFO_SIZE 256
+#define FIFO_NAME SendCarCAN_Q
+#include "fifo.h"
+
+// Motor message
+#define NUM_MOTOR_MSGS 15 // Messages received from the motor controller
+#define MOTOR_MSG_BASE_ADDRESS 0x240
+
+static SendCarCAN_Q_t CANFifo; 
+
+static OS_SEM CarCAN_Sem4;
+static OS_MUTEX CarCAN_Mtx;
+
+// Counter array to reduce the frequency at which we send out messages
+static uint8_t motorMsgCount[NUM_MOTOR_MSGS] = {0};
+static uint8_t motorMsgThreshold[NUM_MOTOR_MSGS] = {5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5};
+
+static void Task_PutIOState(void *p_arg);
 
 /**
- * @brief Sends the feedback from the motor controller to be read by telemetry and 
- * sends pedal, switch, light, and contactor information to be read by telemetry
- * 
- * @param p_arg 
- */
+ * @brief print SendCarCAN_Q put/get for debug purposes
+*/
+void print_SendCarCAN_Q(void) {
+    if(SendCarCAN_Q_is_full(&CANFifo)) {
+        printf("\n\rSendCarCAN_Q is full.");
+    }
+    printf("\n\rSendCarCAN_Q put: %d\n\rSendCarCAN_Q get: %d", CANFifo.put, CANFifo.get);
+}
+
+/**
+ * @brief Wrapper to put new message in the CAN queue
+*/
+void SendCarCAN_Put(CANDATA_t message){
+    OS_ERR err;
+    CPU_TS ticks;
+
+    // Reduce the frequency at which we forward motor messages to CarCAN
+    uint8_t msgIdx = message.ID - MOTOR_MSG_BASE_ADDRESS;
+
+    if (msgIdx >= 0 && msgIdx < NUM_MOTOR_MSGS) {  // Check if the message is from the motor controller
+        if (++motorMsgCount[msgIdx] < motorMsgThreshold[msgIdx]) {
+            return;  
+        } else {
+            motorMsgCount[msgIdx] = 0; // Reset counter and continue sending
+        }     
+    }
+    
+    OSMutexPend(&CarCAN_Mtx, 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(err);
+    
+    bool success = SendCarCAN_Q_put(&CANFifo, message);
+
+    OSMutexPost(&CarCAN_Mtx, OS_OPT_POST_NONE, &err);
+    assertOSError(err);
+
+    if(success) OSSemPost(&CarCAN_Sem4, OS_OPT_POST_1, &err);
+    assertOSError(err);
+}
+
+/**
+ * @brief Initialize SendCarCAN
+*/
+void SendCarCAN_Init(void) {
+    OS_ERR err;
+    
+    OSMutexCreate(&CarCAN_Mtx, "CarCAN_Mtx", &err);
+    assertOSError(err);
+    
+    OSSemCreate(&CarCAN_Sem4, "CarCAN_Sem4", 0, &err);
+    assertOSError(err);
+
+    SendCarCAN_Q_renew(&CANFifo);
+}
+
+/**
+ * @brief Grabs the latest messages from the queue and sends over CarCAN
+*/
 void Task_SendCarCAN(void *p_arg){
-    CANDATA_t motorMsg;
     OS_ERR err;
+    CPU_TS ticks;
+
+    CANDATA_t message;
+    memset(&message, 0, sizeof(CANDATA_t));
```

Comment:
A nice way to do this (which is also type-agnostic) is `memset(&message, 0, sizeof message);`